Add CLI for route flow counter

Junchao-Mellanox opened the pull request:
<!--
    Please make sure you've read and understood our contributing guidelines:
    https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

    ** Make sure all your commits include a signature generated with `git commit -s` **

    If this is a bug fix, make sure your description includes "closes #xxxx",
    "fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
    issue when the PR is merged.

    If you are adding/modifying/removing any command or utility script, please also
    make sure to add/modify/remove any unit tests from the tests
    directory as appropriate.

    If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
    subcommand, or you are adding a new subcommand, please make sure you also
    update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
    your changes.

    Please provide the following information:
-->

#### What I did

#### How I did it

#### How to verify it

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)

Junchao-Mellanox committed:
Add CLI for route flow counter

dgsudharsan committed:
[config reload] Fixing config reload when timer based delayed services are disabled (#1967)

#### What I did
When timer based delayed services like mgmt-framework, telemetry and snmp are disabled and config reload is execute it fails Failed to reset failed state of unit mgmt-framework.service: Unit mgmt-framework.service not loaded.
The reason is these services don't get masked like regular services and these are derived from timers. So when reset-failed is tried on these services it leads to exception.


#### How I did it
When the feature related to these services are disabled their timers would be masked and wouldn't be "enabled". So when deriving the services from timers the services which are not enabled will be skipped.


#### How to verify it
Disable services like mgmt-framework, snmp and telemetry and execute config reload. The config reload should execute without failure

stephenxs committed:
Support CLI for buffer queue configuration (#1965)

What I did
Add CLI for configuring buffer profiles for queues

How to verify it
Run unit test to verify the logic

New command output (if the output of a command-line utility has changed)
config interface buffer queue add
config interface buffer queue set
config interface buffer queue remove

Signed-off-by: Stephen Sun <stephens@nvidia.com>

Junchao-Mellanox committed:
[flow counter] Fix issue: should not compare str with int (#2001)

*When clearing stats, current flow counter stats is serialized to a file for further compare. All stats value is saved as string value. When doing diff, the loaded stats value shall be convert into int and do the compare. The PR is to fix this issue.

ghooo committed:
[GCU] Show default option for '--format' (#2003)

#### What I did
Show the default enum for the option `--format`

#### How I did it
Added `show_default=True`

#### How to verify it
Checkout output below

#### Previous command output (if the output of a command-line utility has changed)
```bash
admin@vlab-01:~$ sudo config apply-patch -h
Usage: config apply-patch [OPTIONS] PATCH_FILE_PATH

  Apply given patch of updates to Config. A patch is a JsonPatch which
  follows rfc6902. This command can be used do partial updates to the config
  with minimum disruption to running processes. It allows addition as well
  as deletion of configs. The patch file represents a diff of ConfigDb(ABNF)
  format or SonicYang format.

  <patch-file-path>: Path to the patch file on the file-system.

Options:
  -f, --format [CONFIGDB|SONICYANG]
                                  format of config of the patch is either
                                  ConfigDb(ABNF) or SonicYang
  -d, --dry-run                   test out the command without affecting
                                  config state
  -v, --verbose                   print additional details of what the
                                  operation is doing
  -h, -?, --help                  Show this message and exit.
admin@vlab-01:~$ 
```

#### New command output (if the output of a command-line utility has changed)
```sh
admin@vlab-01:~$ sudo config apply-patch -h
Usage: config apply-patch [OPTIONS] PATCH_FILE_PATH

  Apply given patch of updates to Config. A patch is a JsonPatch which
  follows rfc6902. This command can be used do partial updates to the config
  with minimum disruption to running processes. It allows addition as well
  as deletion of configs. The patch file represents a diff of ConfigDb(ABNF)
  format or SonicYang format.

  <patch-file-path>: Path to the patch file on the file-system.

Options:
  -f, --format [CONFIGDB|SONICYANG]
                                  format of config of the patch is either
                                  ConfigDb(ABNF) or SonicYang  [default:
                                  CONFIGDB]
  -d, --dry-run                   test out the command without affecting
                                  config state
  -v, --verbose                   print additional details of what the
                                  operation is doing
  -h, -?, --help                  Show this message and exit.
admin@vlab-01:~$ 
```

ghooo committed:
[GCU] Disallowing DeleteInsteadOfReplaceMoveExtender from generating delete whole config move (#2006)

#### What I did
Not generating delete whole config move because it is not allowed by JsonPatch library and it is not possible for ConfigDb to be equal to NULL.

This solves the first problem in issue https://github.com/Azure/sonic-utilities/issues/2000#issuecomment-1011537050

#### How I did it
- Verified UpperLevelMoveExtender does not generate delete whole config move
- Modified DeleteInsteadOfReplaceMoveExtender to not generate delete whole config move
- DeleteRefsMoveExtender cannot generate delete whole config move, because it deletes the referrer leafs

#### How to verify it
unit-test

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)

Commit:
[Barefoot] Added CLI to list/set P4 profile (#1951)

#### What I did
* Added "show platform barefoot profile" command
* Added "config platform barefoot profile" command
* Updated Command Line Reference Guide

#### How I did it
* Created plugins for Barefoot in show/plugins/barefoot.py and config/plugins/barefoot.py
* Updated doc/Command-Reference.md

ghooo committed:
[build-failure-fix][GCU] Sort referrer paths alphabetically

ghooo committed:
Revert "[Barefoot] Added CLI to list/set P4 profile (#1951)"

This reverts commit fa23bfc07d6b2e0fd04e7ab9bba7741af3179f00.

Junchao-Mellanox commented on clear/main.py:
```diff
@@ -490,6 +492,50 @@ def flowcnt_trap():
+    command = "flow_counters_stat -c -t route --prefix_pattern {}".format(prefix_pattern)
+    if vrf:
+        command += ' --vrf {}'.format(vrf)
+    if namespace is not None:
```
add a comment here why don't use "if namespace"

Junchao-Mellanox commented on config/flow_counters.py:
```diff
@@ -0,0 +1,138 @@
+            _, existing_prefix = extract_route_pattern(existing_key)
+        if not existing_prefix:
+            click.echo('Detect invalid route pattern in existing configuration {}, removing it...'.format(existing_key))
+            cfgdb.set_entry(FLOW_COUNTER_ROUTE_PATTERN_TABLE, existing_key, None)
```
ask user if need remove the existing invalid configuration.

Junchao-Mellanox commented on config/flow_counters.py:
```diff
@@ -0,0 +1,138 @@
+    :return:
+    """
+    check_route_flow_counter_support()
+    try:
```
don't check this while removing

Junchao-Mellanox commented on flow_counter_util/route.py:
```diff
@@ -0,0 +1,80 @@
+
+    if support != 'true':
+        print('Route flow counter is not supported on this platform')
+        exit(0)
```
exit(1)?

Junchao-Mellanox commented on scripts/flow_counters_stat:
```diff
@@ -115,6 +121,8 @@ class FlowCounterStats(object):
             table (list): Table data
         """
         if self.args.json:
+            # The first column of the table might have duplicate value, have to
```
remove unused comment

xumia committed:
Fix show ip bgp nei command rw required issue (#2011)

Fix the ro command rw permission required issue

stepanblyschak committed:
Revert "Revert "[sonic-package-manager] support sonic-cli-gen and packages with YANG model (#1650)" (#1972)" (#1994)

This reverts commit fe00bbf63283ca6f3842e470b88cdc5b6d266a0d.

- What I did
Revert previous revert, since the proposed fix has been merged - Azure/sonic-buildimage#9587

- How I did it
Revert the revert.

- How to verify it
Run build an on the switch.

keboliu commented on config/flow_counters.py:
```diff
@@ -0,0 +1,154 @@
+                entry_data = cfgdb.get_entry(FLOW_COUNTER_ROUTE_PATTERN_TABLE, key)
+                old_max_allowed_match = entry_data.get(FLOW_COUNTER_ROUTE_MAX_MATCH_FIELD)
+                if old_max_allowed_match is not None and int(old_max_allowed_match) == max_allowed_match:
+                    click.echo('The route pattern is already configured, no configuration changed')
```
maybe can rephrase like: 'The route pattern already exists, nothing to be changed.'

ghooo committed:
Revert "Revert "[Barefoot] Added CLI to list/set P4 profile (#1951)"" (#2019)

Refer to #1951 for details

The PR #1951 was reverted in #2016 because it was thought to be causing the build failure in sonic-swss TestWarmReboot UTs. But it seems the failures in TestWarmReboot are still occurring e.g. #2017 [build](https://dev.azure.com/mssonic/build/_build/results?buildId=65653&view=logs&s=859b8d9a-8fd6-5a5c-6f5e-f84f1990894e)

I think we can reapply #1951 and need to investigate TestWarmReboot tests instability

dgsudharsan committed:
[fdbshow] Adding more options for fdbshow and show mac (#1982)

#### What I did
 Added more options to filter output in show mac and fdbshow command.
Introduced options for filter by address and filter by type.
Added one more option to display only count.
Introduced show command to display fdb aging time in the switch.

#### How I did it
Modifying fdbshow and show scripts to include the above-mentioned options

#### How to verify it
Added UT for all the newly introduced options and commands

#### Previous command output (if the output of a command-line utility has changed)
```
show mac -h
Usage: show mac [OPTIONS]

  Show MAC (FDB) entries

Options:
  -v, --vlan TEXT
  -p, --port TEXT
  --verbose        Enable verbose output
  -h, -?, --help   Show this message and exit.
```
#### New command output (if the output of a command-line utility has changed)
```
show mac -h
Usage: show mac [OPTIONS] COMMAND [ARGS]...

  Show MAC (FDB) entries

Options:
  -v, --vlan TEXT
  -p, --port TEXT
  -a, --address TEXT
  -t, --type TEXT
  -c, --count
  --verbose           Enable verbose output
  -h, -?, --help      Show this message and exit.

Commands:
  aging-time

show mac
  No.    Vlan  MacAddress         Port         Type
-----  ------  -----------------  -----------  -------
    1      10  98:03:9B:82:BB:5B  Ethernet60   Dynamic
    2      10  EC:0D:9A:CD:91:72  Ethernet64   Dynamic
    3      10  EC:0D:9A:CD:91:73  Ethernet124  Dynamic
Total number of entries 3

show mac --address EC:0D:9A:CD:91:72
  No.    Vlan  MacAddress         Port        Type
-----  ------  -----------------  ----------  -------
    1      10  EC:0D:9A:CD:91:72  Ethernet64  Dynamic

show mac --count
Total number of entries 3

show mac --type Dynamic
  No.    Vlan  MacAddress         Port         Type
-----  ------  -----------------  -----------  -------
    1      10  98:03:9B:82:BB:5B  Ethernet60   Dynamic
    2      10  EC:0D:9A:CD:91:72  Ethernet64   Dynamic
    3      10  EC:0D:9A:CD:91:73  Ethernet124  Dynamic
Total number of entries 3

show mac aging-time
Aging time for switch is 600 seconds
```

vivekrnv committed:
[Auto Techsupport] Added Event Driven TS to Command Reference  (#1985)

Added the Event Driven Tech Support related information to the Command Reference Guide. HLD

vivekrnv committed:
[debug dump] Missing Dict Key handled in the MatchOptimizer (#2014)

- Why I did
The hardcoded delimiter used in the match_optimizer was updated to be assigned dynamically based on the db name provided.
Sometimes, the kv-pairs received from match engine may not contain the necessary fields requested and in those cases MatchEngine Optimizer is throwing a KeyError Exception on the missing field. Added Unit tests to cover the scenario

Traceback (most recent call last):
  File "/usr/local/bin/dump", line 8, in <module>
    sys.exit(dump())
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/dump/main.py", line 85, in state
    collected_info[arg] = obj.execute(params)
  File "/usr/local/lib/python3.9/dist-packages/dump/plugins/route.py", line 82, in execute
    self.init_asic_nh()
  File "/usr/local/lib/python3.9/dist-packages/dump/plugins/route.py", line 142, in init_asic_nh
    nh_ex.collect()
  File "/usr/local/lib/python3.9/dist-packages/dump/plugins/route.py", line 211, in collect
    rif_oid = self.init_asic_next_hop_info(self.rt.nh_id)
  File "/usr/local/lib/python3.9/dist-packages/dump/plugins/route.py", line 191, in init_asic_next_hop_info
    ret = self.rt.nhgrp_match_engine.fetch(req)
  File "/usr/local/lib/python3.9/dist-packages/dump/match_infra.py", line 430, in fetch
    return self.__mutate_response(ret, fv_requested, ret_just_keys)
  File "/usr/local/lib/python3.9/dist-packages/dump/match_infra.py", line 389, in __mutate_response
    new_ret["return_values"][key][field] = key_fv[key][field]
KeyError: 'SAI_NEXT_HOP_ATTR_ROUTER_INTERFACE_ID'

- How I did it
Handled the no-field present case gracefully

Signed-off-by: vkarri <vkarri@contoso.com>

renukamanavalan committed:
[generic-config-updater] Handle failed service restarts (#2020)

What I did
During config update, update of certain tables do demand service restart.
With multiple related updates are not grouped together, this might result in too many service restarts, which could fail with "hitting start limit". When that happens, call reset-failed, try to restart. If it fails again, take a pause and try to restart again.

How I did it
When service restart fails, call reset-failed, try, pause and then call service restart again.

renukamanavalan committed:
[generic_config_updater] Minor update - No logical code change (#2028)

What I did
Missed update from review comments in PR #2020
s/os.system("sleep 10s")/time.sleep(10)/
